### PR TITLE
Synth and onboard speaker support

### PR DESCRIPTION
**Pull request by riknoll**
TODO:

* Sometimes sound cuts out
* Some APIs should not appear if there isn't a built-in speaker

### CODE REVIEW

File: libs/music/music.cpp [hunk 1]
```diff
@@ -0,0 +1,106 @@
+#include "pxt.h"
+
+#include "SAMD21DAC.h"
+#include "Synthesizer.h"
+#include "DeviceConfig.h"
+
+#ifndef HAS_SPEAKER
+#define HAS_SPEAKER PIN_SPEAKER != NC
```
Comment (pelikhan) [inline libs/music/music.cpp]: Parenthesis to be on the safe side?

File: libs/music/music.cpp [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+            synth->setFrequency((float) frequency);
+
+            fiber_sleep(ms);
+            synth->setVolume(0);
+            wait_ms(20);
+        }
+    }
+    else {
+        PwmPin pitchPin = lookupPin(PIN_A10);
```
Comment (pelikhan) [inline libs/music/music.cpp]: Speaker pin?

File: libs/music/music.cpp [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+/**
+* Sets the output volume of the on board speaker (if available)
+* @param volume the volume 0...1023
+*/
+//% weight=96
+//% blockId=synth_set_volume block="set speaker volume %volume"
+//% parts="speaker" blockGap=8 advanced=true
+void setSpeakerVolume(int volume) {
+    synthVolume = volume;
```
Comment (pelikhan) [inline libs/music/music.cpp]: Shouldn't you set the volume now?

File: libs/music/music.cpp [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+//% help=music/play-tone weight=90
+//% blockId=music_play_note block="play tone|at %note=device_note|for %duration=device_beat" blockGap=8
+//% parts="headphone" async
+//% blockNamespace=music
+void playTone(int frequency, int ms) {
+    if (current == SoundOutputDestination::Speaker) {
+        auto synth = &getWSynthesizer()->synth;
+
+        if (frequency <= 0) {
+            fiber_sleep(ms);
```
Comment (pelikhan) [inline libs/music/music.cpp]: set volume to 0

File: libs/music/music.cpp [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        }
+        else {
+            synth->setVolume(synthVolume);
+            wait_ms(20);
+
+            synth->setFrequency((float) frequency);
+
+            fiber_sleep(ms);
```
Comment (pelikhan) [inline libs/music/music.cpp]: if ms <= 0, leave it on

File: libs/music/music.cpp (hunk 2 above)
Comment (riknoll) [inline libs/music/music.cpp]: Nope. This is for when the destination is set to "pin" instead of "speaker"

File: libs/music/music.cpp (hunk 5 above)
Comment (riknoll) [inline libs/music/music.cpp]: why?

File: libs/music/music.cpp (hunk 3 above)
Comment (riknoll) [inline libs/music/music.cpp]: No, that would cause the synth to start making noise.

Comment (riknoll) [inline libs/music/music.cpp]: Hmmm... maybe I can check if the volume isn't equal to 0

Comment (riknoll) [inline libs/music/music.cpp]: Doesn't seem like that is exposed in CODAL.

File: libs/music/music.cpp [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    current = out;
+}
+
+/**
+* Sets the output volume of the on board speaker (if available)
+* @param volume the volume 0...1023
+*/
+//% weight=96
+//% blockId=synth_set_volume block="set speaker volume %volume"
+//% parts="speaker" blockGap=8 advanced=true
```
Comment (pelikhan) [inline libs/music/music.cpp]: specify .min .max bounds to get slider

File: libs/music/music.cpp (hunk 3 above)
Comment (pelikhan) [inline libs/music/music.cpp]: Feature request for Joe or add it yourself

File: libs/music/music.cpp [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        else {
+            synth->setVolume(synthVolume);
+            wait_ms(20);
+
+            synth->setFrequency((float) frequency);
+
+            if (ms > 0) {
+                fiber_sleep(ms);
+                synth->setVolume(0);
+                wait_ms(20);
```
Comment (pelikhan) [inline libs/music/music.cpp]: pin path has 5ms

File: libs/music/music.cpp [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+    if (current == SoundOutputDestination::Speaker) {
+        auto synth = &getWSynthesizer()->synth;
+
+        if (frequency <= 0) {
+            synth->setVolume(0);
+            fiber_sleep(ms);
+        }
+        else {
+            synth->setVolume(synthVolume);
+            wait_ms(20);
```
Comment (pelikhan) [inline libs/music/music.cpp]: is this needed?

**Pull request merged**